Mathlib/Probability/Hoeffding.lean

```diff
@@ -0,0 +1,233 @@
+/-
+Copyright (c) 2024 Kei Tsukamoto. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kei Tsukamoto, Kazumi Kasaura, Yuma Mizuno, Naoto Onda, Sho Sonoda
+-/
+
+import Mathlib.Probability.Moments.Basic
+import Mathlib.Probability.Moments.MGFAnalytic
+import Mathlib.Probability.Variance
+import Mathlib.Analysis.Calculus.ParametricIntegral
+import Mathlib.Analysis.SpecialFunctions.Log.Deriv
+import Mathlib.MeasureTheory.Measure.Tilted
+import Mathlib.Analysis.Calculus.Taylor
+
+/-!
+# Hoeffding's lemma
+
+This file states Hoeffding's lemma.
+
+## Main results
+
+* `ProbabilityTheory.hoeffding`: Hoeffding's Lemma states that for a random variable `X` with
+  `E[X] = 0` (zero mean) and `a ≤ X ≤ b` almost surely, the inequality
+  `mgf X μ t ≤ exp (t^2 * (b - a)^2 / 8)` holds almost surely for all `t ∈ ℝ`.
+
+## References
+
+We follow [martin2019] and [mehryar2018] for the proof of Hoeffding's lemma.
+-/
+
+open MeasureTheory ProbabilityTheory Real
+
+namespace ProbabilityTheory
+
+universe u
+
+variable {Ω : Type u} [MeasurableSpace Ω] (μ : Measure Ω := by volume_tac)
+
+lemma integrable_exp_of_ae_le_const [IsFiniteMeasure μ] {X : Ω → ℝ} (b : ℝ)
+    (hX : AEMeasurable X μ) (hb : ∀ᵐ ω ∂μ, X ω ≤ b) :
+    ∀ (t : ℝ), 0 ≤ t → t ∈ integrableExpSet X μ := by
+  intro t ht
+  refine .of_mem_Icc 0 (rexp (t * b)) (measurable_exp.comp_aemeasurable (hX.const_mul t)) ?_
+  filter_upwards [hb] with ω hb using ⟨by positivity, by gcongr⟩
+
+lemma integrable_exp_of_ae_const_le [IsFiniteMeasure μ] {X : Ω → ℝ} (a : ℝ)
+    (hX : AEMeasurable X μ) (hb : ∀ᵐ ω ∂μ, a ≤ X ω) :
+    ∀ (t : ℝ), t ≤ 0 → t ∈ integrableExpSet X μ := by
+  intro t ht
+  dsimp [integrableExpSet]
+  rw [(by funext ω; simp only [mul_neg, neg_mul, neg_neg]:
+    (fun ω ↦ rexp (t * X ω)) = (fun ω ↦ rexp (-t * -X ω)))]
+  apply integrable_exp_of_ae_le_const
+  · exact AEMeasurable.neg hX
+  · filter_upwards [hb] with ω hb using neg_le_neg_iff.mpr hb
+  · exact neg_nonneg.mpr ht
+
+lemma integrable_exp_of_ae_mem_Icc [IsFiniteMeasure μ] {X : Ω → ℝ} (a b : ℝ)
+    (hX : AEMeasurable X μ) (hb : ∀ᵐ ω ∂μ, X ω ∈ Set.Icc a b) :
+    integrableExpSet X μ = Set.univ := by
+  apply Set.eq_univ_iff_forall.mpr
+  intro t
+  by_cases sign_t : 0 ≤ t
+  case pos =>
+    apply integrable_exp_of_ae_le_const _ _ hX
+      (by filter_upwards [hb] with ω hb using hb.2) _ sign_t
+  case neg =>
+    apply integrable_exp_of_ae_const_le _ _  hX
+      (by filter_upwards [hb] with ω hb using hb.1) _ (le_of_not_ge sign_t)
+
+theorem integrable_exp_set_interior_of_ae_mem_Icc  [IsFiniteMeasure μ] {X : Ω → ℝ} (t a b : ℝ)
+    (hX : AEMeasurable X μ)
+    (h : ∀ᵐ (ω : Ω) ∂μ, X ω ∈ Set.Icc a b) : t ∈ interior (integrableExpSet X μ) := by
+  rw [integrable_exp_of_ae_mem_Icc μ a b hX h]
+  apply mem_interior_iff_mem_nhds.mpr Filter.univ_mem
+
+theorem cgf_deriv_eq_tilted_measure_expectation {X : Ω → ℝ} (t : ℝ)
```

Comment:
```suggestion
theorem deriv_cgf_eq_integral_tilted {X : Ω → ℝ} (t : ℝ)
```

Reply:
This lemma and the next one are now in Mathlib, in `Mathlib.Probability.Moments.Tilted` (in a more general version for the one on the variance, without any bounded assumption).